packages/flutter/lib/src/widgets/route_name.dart

```diff
@@ -0,0 +1,67 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/rendering.dart';
+import 'package:flutter/widgets.dart';
+
+/// A widget which provides a semantic name for a route.
+///
+/// On Android, TalkBack uses [routeName] as the text value for an edge 
+/// triggered semantics update. On iOS, no additional semantic information
+/// is inserted since by convention VoiceOver uses a standard chime sound 
+/// effect for most navigation events.
+///
+/// See also:
+///
+///  * [SemanticsProperties.namesRoute], for a description of how route name
+///    semantics work.
+class RouteName extends StatelessWidget  {
+  /// Creates a widget which provides a semantic route name.
+  ///
+  /// [child] and [routeName] are required arguments.
+  const RouteName({
+    Key key,
+    @required this.child,
+    @required this.routeName,
+  }) : assert(child != null),
+       super(key: key);
+
+  /// A semantic label for the route.
+  /// 
+  /// On iOS platforms this value is ignored.
+  final String routeName;
+
+  /// The widget below this widget in the tree.
+  ///
+  /// {@macro flutter.widgets.child}
+  final Widget child;
+
+  @override
+  Widget build(BuildContext context) {
+    Widget result;
+    switch (defaultTargetPlatform) {
+      case TargetPlatform.iOS:
+        result = child;
+        break;
+      case TargetPlatform.android:
+      case TargetPlatform.fuchsia:
+        result = new Semantics(
+          scopesRoute: true,
```

Comment:
It also scopes a route? The doc above made it sound like it only gives a name to a surrounding route?

Reply:
Also: If I now use this within a MaterialRoute to give the route a custom name wouldn't there be two scopeRoutes in the tree for the same route?

Reply:
Having said this, if this doesn't specify scopesRoute, how would we ensure that this node doesn't become focusable?

Reply:
The "last" scopes route flag is selected, so this route name will take precedence.  This allows users to name custom routes and makes this useful for dialogs and drawers too

Reply:
Also update doc

Reply:
But wouldn't the semantic tree make it appear as if there are two nested routes on screen? I know it's fine for our current platforms, but what if other platforms decide to announce every route on screen?

Reply:
I think that either approach is okay for now, as it won't impact users building a standard cupertino/material style application.  But I also agree it would be best not to introduce redundant semantic information.

In the case of custom navigator/pages, having the RouteName also scope a route _might_ make it slightly easier to provide a working implementation.  But again, the implementation of namesRoute means that we don't have to.

Having them scope a route also let's me reuse the widget for Drawer, Dialog, and PopupMenu - but since we maintain these widgets I don't think it is a big deal for us to just directly add the correct semantic flags there.

@Hixie originally proposed using the route name for the other widgets, I'm not sure if he had a different idea from what I cahnged

Reply:
Oh right, the other issue is that the `RouteName` widget adds semantic information but doesn't really correspond to a focusable element, and it probably shouldn't just get merged with one.  And since we don't control focus on the framework side, I can  piggyback on the focus logic I added for scopesRoute to make sure it is ignored correctly